### PR TITLE
SegWit - witness script templates

### PR DESCRIPTION
**Pull request by dcousens**
A derivative of the preparatory work done by @rubensayshi in https://github.com/bitcoinjs/bitcoinjs-lib/pull/520.
This purely encompasses the `script` changes and adds relevant tests for them.

This change should be atomic enough such that it can be merged easily.
~~Before merge, an integration test showing the full process of steps to inter-operate with SegWit should be added.~~

### CODE REVIEW

**Comment by rubensayshi** [comment]
are you gonna have more for the transaction de/seralializing and signing? or should I clean up my PR and rebase on this?

File: src/script.js [hunk 1]
```diff
@@ -206,6 +206,20 @@ function isScriptHashOutput (script) {
     buffer[22] === OPS.OP_EQUAL
 }
 
+function isWitnessPubKeyHashOutput (script) {
+  var buffer = compile(script)
+
+  return buffer.length === 22 &&
+    buffer[0] === OPS.OP_0
```
Comment (afk11) [inline src/script.js]:
```
+    buffer[1] === 0x14 
```
is worth checking in addition to checking the buffer size, so you know the data push is actually 20 bytes

File: src/script.js [hunk 2]
```diff
@@ -206,6 +206,20 @@ function isScriptHashOutput (script) {
...
+}
+
+function isWitnessScriptHashOutput (script) {
+  var buffer = compile(script)
+
+  return buffer.length === 34 &&
+    buffer[0] === OPS.OP_0
```
Comment (afk11) [inline src/script.js]:
```
+    buffer[1] === 0x20
```

**Comment by dcousens** [comment]
> are you gonna have more for the transaction de/seralializing and signing?

I was going to keep it atomic,  but I guess for the proposed integration test, I'll need to.

No `TransactionBuilder` changes at this time though.

**Comment by dcousens** [comment]
@afk11 resolved your comments in b02761d.  Great suggestions :)

File: src/script.js [hunk 3]
```diff
@@ -383,14 +422,21 @@ module.exports = {
...
   isNullDataOutput: isNullDataOutput,
+
   classifyOutput: classifyOutput,
   classifyInput: classifyInput,
   pubKeyOutput: pubKeyOutput,
   pubKeyHashOutput: pubKeyHashOutput,
   scriptHashOutput: scriptHashOutput,
+  witnessPubKeyHashOutput: witnessPubKeyHashOutput,
+  witnessScriptHashInput: witnessScriptHashInput,
+  witnessScriptHashOutput: witnessScriptHashOutput,
```
Comment (dcousens) [inline src/script.js]: @rubensayshi I also opted for `witness*` such that it conformed to the standard we had used elsewhere,  which was to follow the P2SH, P2PKH, P2WPKH, P2WSH notation,  but,  without the `P2`.

**Comment by rubensayshi** [comment]
> No TransactionBuilder changes at this time though.

so you don't want bitcoinjs-lib to be able to sign segwit?  
afaik my PR doesn't BC break anything, it just touches quite a few things internally

**Comment by dcousens** [comment]
@rubensayshi I'm just doing it in small byte sized PRs.  The `TransactionBuilder` changes are going to require more thought IMHO

**Comment by dcousens** [comment]
@rubensayshi like,   I think we can do this better (solutions welcome)
``` js
TransactionBuilder.prototype.sign = function (index, keyPair, redeemScript, hashType, segWit, amount, segWitScript)
```

**Comment by dcousens** [comment]
Perhaps for now,  lets not add the `Transaction` code to this PR yet,  and just merge if we're all happy with this interface.

**Comment by dcousens** [comment]
@rubensayshi @jprichardson please ACK if you're happy with this interface, and we'll then look to rebasing the other PR such that we can iterate the `Transaction` and `TransactionBuilder` API's

**Comment by rubensayshi** [comment]
ACK

**Comment by jprichardson** [comment]
ACK

**Pull request merged**